File: cpp/Data Structures/SegmentTree Iterativo.cpp


// [1, n]
template<typename T>
struct STree {
    vector<T> tree;
    int n;
    T neutro = T(0);

    T oper(T a, T b) {
        return a + b;
    }

    STree(vector<T> &a) {
        int tam = 1;
        while(1LL << (tam) < sz(a)) tam++;
        n = 1LL << tam;
        tree.resize(n * 2);
        forab(i, 1, sz(a) + 1) upd(i, a[i - 1]);
    }

    void upd(int i, T x) {
        i += n;
        tree[i] = x;
        while (i > 1) {
            i /= 2;
            // The index of the left child is s * 2.
            // The index of the right child is s * 2 + 1.
            tree[i] = oper(tree[i * 2], tree[i * 2 + 1]);
        }
    }

    T qry(int l, int r) {
        T res = 0;
        l += n;
        r += n;
        while (l <= r) {
            if (l % 2 == 1) res = oper(res, tree[l++]);
            if (r % 2 == 0) res = oper(res, tree[r--]);
            l /= 2;
            r /= 2;
        }
        return res;
    }

    //lower bound in the prefix sum
    int lb(T k) {
        int s = 1;
        while (s < n) {
            if (tree[s * 2] >= k) {
                s = s * 2;
            } else {
                k -= tree[s * 2];
                s = s * 2 + 1;
            }
        }

        return s - n;
    }
};

File: Java/Data Structures/Segment Tree Iterativo.java
static class Node{
    long val;
    Node(long v){
        val = v;
    }
}
static class STree {
    int n;
    Node neutro = new Node(0);
    Node st[];
    STree(int N) {
        n = N;
        for (int i = 0; i < (1<<n); i++)
            st[i] = neutro;
    }
    STree(long a[]) {
        n = a.length;
        st = new Node[1 << n];
        build(a);
    }
    Node oper(Node a, Node b) {
        return new Node(a.val + b.val);
    }
    void build(long a[]) {
        for (int i = 0; i < n; i++)
            st[n + i] = new Node(a[i]);
        for (int i = n - 1; i >= 1; i--)
            st[i] = oper(st[i << 1], st[i << 1 | 1]);
    }
    Node query(int l, int r) { // [l, r]
        Node vl = neutro, vr = neutro;
        for (l += n, r += n + 1; l < r; l >>= 1, r >>= 1) {
            if ((l & 1) == 1) vl = oper(vl, st[l++]);
            if ((r & 1) == 1) vr = oper(st[--r], vr);
        }
        return oper(vl, vr);
    }
    void upd(int p, long val) { // set val at position p (0 - idx)
        for (st[p += n] = new Node(val); p > 1; p >>= 1)
            st[p >> 1] = oper(st[p], st[p ^ 1]);
    }
}